### PR TITLE
Creating a FleetAllocation allocated a GameServer from a Fleet

### PR DESCRIPTION
**Pull request by markmandel**
This implements the FleetAllocation CRD, that when created looks at the requested Fleet, and marks one of the GameServers as Allocated (if available) and returns it's value in the FleetAllocation.

This also includes a quickstart and spec for reference as well.

Some refactoring was included in the PR (probably should have been a seperate PR for that).

### CODE REVIEW

File: pkg/fleets/controller.go [hunk 1]
```diff
@@ -169,7 +168,7 @@ func (c *Controller) syncFleet(key string) error {
 		return errors.Wrapf(err, "error retrieving fleet %s from namespace %s", name, namespace)
 	}
 
-	list, err := c.listGameServerSets(fleet)
+	list, err := ListGameServerSetsByFleetOwner(c.gameServerSetLister, fleet)
```
Comment (markmandel) [inline pkg/fleets/controller.go]: @enocom would love to get your opinion on this refactor ([implementation here](https://github.com/GoogleCloudPlatform/agones/pull/193/files#diff-91029c2ecd86ab36712b39724f333a94R28))

There were several places I needed a list type function like this, and wasn't quite sure the best way to do it.

File: pkg/apis/stable/v1alpha1/fleetallocation.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+// FleetAllocationStatus will contain the
+// `GameServer` that has been allocated from
+// a Fleet
+type FleetAllocationStatus struct {
+	GameServer *GameServer
+}
+
+// Validateupdate validates when an update occurs
+func (fa *FleetAllocation) Validateupdate(new *FleetAllocation) (bool, []metav1.StatusCause) {
```
Comment (EricFortin) [inline pkg/apis/stable/v1alpha1/fleetallocation.go]: Nitpick: ValidateUpdate

**Comment by EricFortin** [comment]
LGTM!

File: pkg/fleets/fleets.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	"agones.dev/agones/pkg/gameserversets"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// ListGameServerSetsByFleetOwner lists all the GameServerSets for a given
+// Fleet
+func ListGameServerSetsByFleetOwner(gameServerSetLister listerv1alpha1.GameServerSetLister, f *stablev1alpha1.Fleet) ([]*stablev1alpha1.GameServerSet, error) {
+	list, err := gameServerSetLister.List(labels.SelectorFromSet(labels.Set{stablev1alpha1.FleetGameServerSetLabel: f.ObjectMeta.Name}))
```
Comment (enocom) [inline pkg/fleets/fleets.go]: Based on a cursory reading of the code, I think you could even replace the concrete type `*stablev1alpha1.Fleet` with a `metav1.Object` interface. This way, the `ListGameServerSetsByFleetOwner` function might enjoy an even wider usage.

Comment (enocom) [reply pkg/fleets/fleets.go]: If you use the `metav1.Object` interface here, then the name lookup becomes `f.GetName()` as with the lookup below.

File: pkg/fleets/controller.go (hunk 1 above)
Comment (enocom) [reply pkg/fleets/controller.go]: See my response below.

File: pkg/fleets/fleets.go (hunk 3 above)
Comment (enocom) [reply pkg/fleets/fleets.go]: I've seen that pattern in some object-oriented code, but I'm having a hard time remembering what the pattern might be called. It seems entirely reasonable to me.

Comment (markmandel) [reply pkg/fleets/fleets.go]: SGTM! Thanks for the sanity check.

Comment (enocom) [reply pkg/fleets/fleets.go]: I think it's basically a form of [double dispatch](https://en.wikipedia.org/wiki/Double_dispatch).

**Pull request merged**